### DISCUSSION TITLE
`brew update` has stopped updating Casks

### DISCUSSION THREAD

**Discussion by owine** [Homebrew/discussions#4134] [Casks]
### Output of `brew config`

```shell
➜  ~ brew config
HOMEBREW_VERSION: 3.6.20-97-g71fab93
ORIGIN: https://github.com/Homebrew/brew
HEAD: 71fab93983b47c73b3a36b24543a1adcea290044
Last commit: 6 minutes ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 2a9101e748417387a8a570304b23ba4b5fa92132
Core tap last commit: 22 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_INSTALL_FROM_API: set
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.39.1 => /opt/homebrew/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
Rosetta 2: false
```

### Output of `brew doctor`

```shell
➜  ~ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  mkvtoolnix
```

### Description of issue

It seems that `brew update` is no longer updating the cask tap on my machine. Merged PRs are not reflected in `brew outdated` or `brew upgrade` and the repo reports it is behind

```
➜  homebrew-cask git:(master) git fetch
➜  homebrew-cask git:(master) git status
On branch master
Your branch is behind 'origin/master' by 285 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)

nothing to commit, working tree clean
➜  homebrew-cask git:(master) brew update
Already up-to-date.
➜  homebrew-cask git:(master) git status
On branch master
Your branch is behind 'origin/master' by 285 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)

nothing to commit, working tree clean
```

### Relevant casks

All

**Comment by SMillerDev** [comment]
Have you tried doing the `git pull` it suggests?

**Reply by owine** [reply]
> Have you tried doing the `git pull` it suggests?

Yes and this operated as expected - no errors.

**Reply by SMillerDev** [reply]
What happens if you set `HOMEBREW_NO_INSTALL_FROM_API=1`

**Reply by owine** [reply]
> What happens if you set `HOMEBREW_NO_INSTALL_FROM_API=1`

When setting this, `brew update` now updates casks as expected.

**Comment by xeron** [comment]
I see similar issue with `brew update` not updating `homebrew/core` and `homebrew/cask` (`git pull` works fine):

```
$ brew update
Already up-to-date.
$ HOMEBREW_NO_INSTALL_FROM_API=1 brew update
Updated 2 taps (homebrew/core and homebrew/cask).
==> New Casks
rive                                                                                                    serene
==> Outdated Formulae
awscli                                   ffmpeg                                   guile                                    imagemagick                              openjdk
==> Outdated Casks
chefdk

You have 5 outdated formulae and 1 outdated cask installed.
You can upgrade them with brew upgrade
or list them with brew outdated.
```

Similar to OP my `HOMEBREW_PREFIX` is not standard (`/opt/brew`). Might be the reason. And before anyone mentions it – no, I'm not willing to use the standard prefix.

**Comment by Rylan12** [comment]
Homebrew is in the process of being updated to install homebrew/core formulae and homebrew/cask casks from the formulae.brew.sh API. When this is done, there's no need for the local git repos to be updated. So, to save time on `brew updates`, we don't update them. If you need to have updated git taps, you'll need to set `HOMEBREW_NO_INSTALL_FROM_API=1` as Sean mentioned.

However, `brew update` and `brew upgrade` should still work as expected even when the taps don't update. Note that it can take up to an hour for updates to become available via the API. Can you confirm which formulae/casks you're expecting to see in `brew outdated`? Also, posting the output of `ls -l "$(brew --cache)/api"` might be helpful in identifying why this isn't working.

**Reply by owine** [reply]
> Homebrew is in the process of being updated to install homebrew/core formulae and homebrew/cask casks from the formulae.brew.sh API. When this is done, there's no need for the local git repos to be updated. So, to save time on `brew updates`, we don't update them. If you need to have updated git taps, you'll need to set `HOMEBREW_NO_INSTALL_FROM_API=1` as Sean mentioned.

Confirmed, this does update the repos.

> However, `brew update` and `brew upgrade` should still work as expected even when the taps don't update. Note that it can take up to an hour for updates to become available via the API. Can you confirm which formulae/casks you're expecting to see in `brew outdated`? Also, posting the output of `ls -l "$(brew --cache)/api"` might be helpful in identifying why this isn't working.

My repos are currently up to date, but I do have a cask that refuses to update, API or not.

```
➜  ~ brew outdated
1password-cli (2.13.0) != 2.13.1
➜  ~ brew upgrade 1password-cli
==> Upgrading 1 outdated package:
Error: Cask '1password-cli' is unavailable: No Cask with this name exists.
➜  ~ HOMEBREW_NO_INSTALL_FROM_API=1 brew upgrade 1password-cli
==> Upgrading 1 outdated package:
Error: Cask '1password-cli' is unavailable: No Cask with this name exists.
```

```
➜  ~ ls -l "$(brew --cache)/api"
total 47360
-rw-r--r--  1 owine  staff   3750194 Jan 28 12:21 cask.json
-rw-r--r--  1 owine  staff  19304736 Jan 28 12:21 formula.json
```

**Reply by xeron** [reply]
What I noticed happening is brew tries to update the formula from the API, e.g. it downloads new version, but then tries to install it using formula from the local repo (outdated) resulting in issues like this:

guile update from 3.0.8_4 to 3.0.9 tried patching 3.0.9 with a patch which has been removed:

```
==> Patching
Error: An exception occurred within a child process:
  Errno::ENOENT: No such file or directory @ rb_sysopen - /Users/…/Library/Caches/Homebrew/downloads/834d775e5d03dc39c416e2ade041ed0e4e8f01cc0b605fc21cef66bf13a34c31--b54263dc98b2700fa777745405ad7651601bcdc6..3bdcc3668fd8f9a5b6c9a313ff8d70acb32b2a52.patch
```

**Reply by xyzzy-plugh-plover** [reply]
I get:

```
$ brew outdated --greedy --verbose 
bettertouchtool (3.9999,2206) != 4.001,2214
google-drive (69.0.0) != 70.0.2
macupdater (2.3.1,14204) != 2.3.2,14210
protonvpn (3.0.12) != 3.0.13

$ export HOMEBREW_NO_INSTALL_FROM_API=1

$ brew upgrade --greedy protonvpn
==> Upgrading 1 outdated package:
Error: Cask 'protonvpn' is unavailable: No Cask with this name exists.

$ brew config
HOMEBREW_VERSION: 3.6.20-167-g7a5f614
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 7a5f6143d3e7447a64255c548d2e51138c83acd3
Last commit: 9 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 7536497f18296a8bd886a067a6695691a4cc322d
Core tap last commit: 25 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.kCgR4AKQiL/org.xquartz:0
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_NO_INSTALL_FROM_API: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit haswell
Clang: 14.0.0 build 1400
Git: 2.39.1 => /usr/local/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.6.3-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2

$ brew doctor
Your system is ready to brew.
```

**Reply by gromgit** [reply]
What's the output of `brew upgrade --debug --verbose --greedy protonvpn`?

**Reply by xyzzy-plugh-plover** [reply]
```
$ brew upgrade --debug --verbose --greedy protonvpn
Running `brew update --auto-update`...
==> Auto-updated Homebrew!
Updated 3 taps (homebrew/cask-versions, homebrew/core and homebrew/cask).
==> New Casks
spotify4bigsur

==> Upgrading 1 outdated package:
Error: Cask 'protonvpn' is unavailable: No Cask with this name exists.
/usr/local/Homebrew/Library/Homebrew/cask/cask_loader.rb:342:in `load'
/usr/local/Homebrew/Library/Homebrew/cask/cask_loader.rb:351:in `load'
/usr/local/Homebrew/Library/Homebrew/cask/cmd/upgrade.rb:149:in `block in upgrade_casks'
/usr/local/Homebrew/Library/Homebrew/cask/cmd/upgrade.rb:149:in `map'
/usr/local/Homebrew/Library/Homebrew/cask/cmd/upgrade.rb:149:in `upgrade_casks'
/usr/local/Homebrew/Library/Homebrew/cmd/upgrade.rb:226:in `upgrade_outdated_casks'
/usr/local/Homebrew/Library/Homebrew/cmd/upgrade.rb:117:in `upgrade'
/usr/local/Homebrew/Library/Homebrew/brew.rb:105:in `<main>'
```

**Reply by gromgit** [reply]
Hmmm. Try `brew reinstall protonvpn`.

**Reply by xyzzy-plugh-plover** [reply]
Thanks, that worked

**Comment by whitslack** [comment]
I'm on High Sierra, an "unsupported" macOS. Running `brew outdated` told me I had no outdated formulae, yet my taps were behind. Doing a `git merge origin/master` in `/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core` fast-forwarded the local checkout, and then `brew outdated` listed several outdated formulae.

**Reply by whitslack** [reply]
This is no longer a problem. `brew update` is fast-forwarding my local checkouts just fine now, even when they contain uncommitted changes.

**Comment by medoasiri** [comment]
https://github.com/orgs/Homebrew/discussions/4134#discussioncomment-4834269
